File: examples/csharp/avalonia/NotesShowcase.Tests/ViewModels/NotesViewVMTests.cs
using System.Reactive.Concurrency;
using Microsoft.Reactive.Testing;
using NotesShowcase.Models;
using NotesShowcase.ViewModels;
using VMx.Lifecycle;
using VMx.Services;
using Xunit;

namespace NotesShowcase.Tests.ViewModels;

public sealed class NotesViewVMTests
{
    private static NotesViewVM BuildVM(
        InMemoryNoteRepository repo,
        IScheduler? searchScheduler = null,
        TimeSpan? debounce = null)
    {
        var hub = new MessageHub();
        var dispatcher = new RxDispatcher(ImmediateScheduler.Instance, ImmediateScheduler.Instance);
        return NotesViewVM.Builder()
            .Name("notes")
            .Services(hub, dispatcher)
            .Repository(repo)
            .PageSize(5)
            .SearchDebounce(debounce ?? TimeSpan.FromMilliseconds(150))
            .SearchScheduler(searchScheduler ?? ImmediateScheduler.Instance)
            .Build();
    }

    private static async Task<(NotesViewVM vm, InMemoryNoteRepository repo)> BuildAndBindAsync(
        string notebookId = "nb-reviews",
        IScheduler? searchScheduler = null,
        TimeSpan? debounce = null)
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            loadNotesDelay: TimeSpan.Zero);
        var vm = BuildVM(repo, searchScheduler, debounce);
        vm.Construct();
        await vm.BindToAsync(notebookId);
        return (vm, repo);
    }

    [Fact]
    public async Task BindToAsync_loads_notes_for_the_notebook()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        Assert.NotEmpty(vm.VisibleItems);
        Assert.All(vm.FilteredItems, n => Assert.Equal("nb-reviews", n.Model.NotebookId));
        Assert.Equal("nb-reviews", vm.BoundNotebookId);
    }

    [Fact]
    public async Task PageSize_5_yields_two_pages_for_seven_reviews_notes()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        // 7 review notes / page-size 5 => 2 pages
        Assert.Equal(2, vm.PageCount);
        Assert.Equal(5, vm.VisibleItems.Count);
        vm.MoveToNextPageCommand.Execute(null);
        Assert.Equal(2, vm.VisibleItems.Count);
    }

    [Fact]
    public async Task Pagination_no_op_at_boundaries()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        var first = vm.CurrentPageIndex;
        Assert.False(vm.MoveToFirstPageCommand.CanExecute(null));
        Assert.False(vm.MoveToPreviousPageCommand.CanExecute(null));
        Assert.True(vm.MoveToNextPageCommand.CanExecute(null));
        Assert.True(vm.MoveToLastPageCommand.CanExecute(null));
        vm.MoveToFirstPageCommand.Execute(null);
        Assert.Equal(first, vm.CurrentPageIndex);
        vm.MoveToLastPageCommand.Execute(null);
        var last = vm.CurrentPageIndex;
        Assert.True(vm.MoveToFirstPageCommand.CanExecute(null));
        Assert.True(vm.MoveToPreviousPageCommand.CanExecute(null));
        Assert.False(vm.MoveToNextPageCommand.CanExecute(null));
        Assert.False(vm.MoveToLastPageCommand.CanExecute(null));
        vm.MoveToNextPageCommand.Execute(null);
        Assert.Equal(last, vm.CurrentPageIndex);
    }

    [Fact]
    public async Task ShowStarredOnly_filter_restricts_to_starred_notes()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        vm.ShowStarredOnly = true;
        Assert.All(vm.FilteredItems, n => Assert.True(n.Starred));
        Assert.NotEmpty(vm.FilteredItems); // nb-reviews has 2 starred per seed
    }

    [Fact]
    public async Task ShowStarredOnly_false_restores_all_notes()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        vm.ShowStarredOnly = true;
        var starredCount = vm.FilteredItems.Count;
        vm.ShowStarredOnly = false;
        Assert.True(vm.FilteredItems.Count > starredCount);
    }

    [Fact]
    public async Task Filter_predicate_restricts_visible_items()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        vm.Filter = n => n.Title.Contains("Q1");
        Assert.Single(vm.FilteredItems);
        Assert.Contains("Q1", vm.FilteredItems[0].Title);
        vm.Filter = null;
        Assert.True(vm.FilteredItems.Count > 1);
    }

    [Fact]
    public async Task BindToAsync_to_different_notebook_swaps_items()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        var beforeIds = vm.FilteredItems.Select(i => i.NoteId).ToList();
        await vm.BindToAsync("nb-work");
        var afterIds = vm.FilteredItems.Select(i => i.NoteId).ToList();
        Assert.NotEqual(beforeIds, afterIds);
        Assert.All(vm.FilteredItems, n => Assert.Equal("nb-work", n.Model.NotebookId));
    }

    [Fact]
    public async Task IsEmpty_true_when_filter_excludes_everything()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        Assert.False(vm.IsEmptyDerived.Value);
        vm.Filter = _ => false;
        Assert.True(vm.IsEmpty);
        Assert.True(vm.IsEmptyDerived.Value);
        Assert.Empty(vm.VisibleItems);
    }

    [Fact]
    public async Task PageLabel_reflects_current_page_and_count()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        Assert.Equal("Page 1 of 2", vm.PageLabel);
        Assert.Equal("Page 1 of 2", vm.PageLabelDerived.Value);
        vm.MoveToNextPageCommand.Execute(null);
        Assert.Equal("Page 2 of 2", vm.PageLabel);
        Assert.Equal("Page 2 of 2", vm.PageLabelDerived.Value);
    }

    [Fact]
    public async Task Search_debounced_via_TestScheduler_only_emits_after_window()
    {
        var ts = new TestScheduler();
        var (vm, _) = await BuildAndBindAsync(
            "nb-reviews",
            searchScheduler: ts,
            debounce: TimeSpan.FromMilliseconds(150));
        // Initial: 7 visible (across 2 pages — 5 on page 1).
        var before = vm.FilteredItems.Count;
        Assert.True(before > 1);

        vm.SearchTerm = "Q1";
        // Before the debounce window elapses, the recompute hasn't fired yet,
        // so FilteredItems still reflects the prior term ("").
        ts.AdvanceBy(TimeSpan.FromMilliseconds(50).Ticks);
        Assert.Equal(before, vm.FilteredItems.Count);
        // Advance past the 150 ms debounce window.
        ts.AdvanceBy(TimeSpan.FromMilliseconds(150).Ticks);
        Assert.Single(vm.FilteredItems);
        Assert.Contains("Q1", vm.FilteredItems[0].Title);
    }

    [Fact]
    public async Task Current_setter_emits_PropertyChanged_and_is_idempotent()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        var first = vm.VisibleItems[0];
        vm.Current = first;
        Assert.Same(first, vm.Current);
        vm.Current = first; // idempotent
        Assert.Same(first, vm.Current);
        vm.Current = null;
        Assert.Null(vm.Current);
    }

    [Fact]
    public async Task Reconstruct_resets_lifecycle_and_keeps_items_via_rebind()
    {
        var (vm, _) = await BuildAndBindAsync("nb-reviews");
        var before = vm.FilteredItems.Count;
        vm.Reconstruct();
        // After reconstruct the prior items were torn down — a rebind restores them.
        await vm.BindToAsync("nb-reviews");
        Assert.Equal(before, vm.FilteredItems.Count);
    }

    // ── end-to-end delete-path coverage: end-to-end delete pathway coverage ────────
    // The full pathway (repo.DeleteNoteAsync → remove from _inner → clear
    // Current → dispose) had 0 % coverage. Drive it through NoteVM's
    // DeleteCommand (the only public entry-point to NotesViewVM.DeleteNote).

    [Fact]
    public async Task DeleteNoteAsync_removes_from_inner_and_clears_current_and_persists()
    {
        var seed = SeedData.Build();
        var repo = new InMemoryNoteRepository(
            seed,
            loadAllDelay: TimeSpan.Zero,
            loadNotesDelay: TimeSpan.Zero,
            saveNoteDelay: TimeSpan.Zero,
            deleteNoteDelay: TimeSpan.Zero);
        var hub = new VMx.Services.MessageHub();
        var dispatcher = new VMx.Services.RxDispatcher(
            System.Reactive.Concurrency.ImmediateScheduler.Instance,
            System.Reactive.Concurrency.ImmediateScheduler.Instance);
        var vm = NotesViewVM.Builder()
            .Name("notes").Services(hub, dispatcher).Repository(repo).PageSize(5)
            .Build();
        vm.Construct();
        await vm.BindToAsync("nb-personal");
        var before = vm.Inner.Count;
        var target = vm.Inner[0];
        vm.Current = target;

        // Drive delete through NoteVM.DeleteCommand (raw, no confirm wrapped
        // for this test build) — this is the path NotesListView.axaml fires
        // when the user clicks the delete button next to a note.
        target.DeleteCommand.Execute(null);
        // Repo delete is awaited inside DeleteNoteAsyncInternal; wait for the
        // continuation to remove the row.
        await TestWait.WaitUntilAsync(() => vm.Inner.Count == before - 1);

        Assert.Equal(before - 1, vm.Inner.Count);
        Assert.Null(vm.Current);
        // Persistence check: the repo no longer returns the deleted note.
        var reload = await repo.LoadNotesAsync("nb-personal");
        Assert.DoesNotContain(reload, n => n.Id == target.NoteId);
    }

    [Fact]
    public async Task Repository_search_notes_returns_token_pages_over_all_notes()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadNotesDelay: TimeSpan.Zero);

        var first = await repo.SearchNotesAsync("review", token: null, pageSize: 2);

        Assert.Equal(2, first.Items.Count);
        Assert.Equal("2", first.NextToken);
        Assert.All(first.Items, n =>
            Assert.Contains("review", $"{n.Title} {n.Body} {string.Join(" ", n.Tags)}".ToLowerInvariant()));

        var second = await repo.SearchNotesAsync("review", first.NextToken, pageSize: 2);

        Assert.NotEmpty(second.Items);
        Assert.NotEqual(first.Items[0].Id, second.Items[0].Id);
    }

    [Fact]
    public async Task Repository_search_notes_rejects_malformed_and_extreme_offsets()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadNotesDelay: TimeSpan.Zero);
        var first = await repo.SearchNotesAsync("review", token: null, pageSize: 2);

        var malformed = await repo.SearchNotesAsync("review", token: "2junk", pageSize: 2);
        var extreme = await repo.SearchNotesAsync(
            "review", token: int.MaxValue.ToString(), pageSize: int.MaxValue);

        Assert.Equal(first.Items, malformed.Items);
        Assert.Empty(extreme.Items);
        Assert.Null(extreme.NextToken);
    }

    [Fact]
    public async Task Direct_dispose_releases_live_note_children_and_binding_state()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadNotesDelay: TimeSpan.Zero);
        var vm = BuildVM(repo);
        vm.Construct();
        await vm.BindToAsync("nb-personal");
        var children = vm.Inner.ToArray();

        vm.Dispose();

        Assert.NotEmpty(children);
        Assert.All(children, child => Assert.Equal(ConstructionStatus.Disposed, child.Status));
        Assert.Empty(vm.Inner);
        Assert.Null(vm.BoundNotebookId);
    }

    [Fact]
    public async Task GlobalSearchVM_refreshes_resets_terms_and_loads_more()
    {
        var hub = new MessageHub();
        var dispatcher = new RxDispatcher(ImmediateScheduler.Instance, ImmediateScheduler.Instance);
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadNotesDelay: TimeSpan.Zero);
        var vm = GlobalSearchVM.Builder()
            .Name("global-search")
            .Services(hub, dispatcher)
            .Repository(repo)
            .PageSize(2)
            .SearchDebounce(TimeSpan.Zero)
            .Build();

        vm.SearchTerm = "review";
        await vm.RefreshCommand.ExecuteAsync();
        Assert.Equal(2, vm.Results.Count);
        Assert.True(vm.HasMore);

        await vm.LoadMoreCommand.ExecuteAsync();
        Assert.True(vm.Results.Count > 2);
        var replacedResults = vm.Results.ToArray();

        vm.SearchTerm = "travel";
        await vm.RefreshCommand.ExecuteAsync();
        Assert.All(vm.Results, n => Assert.Equal("nb-personal", n.Model.NotebookId));
        var finalResults = vm.Results.ToArray();
        vm.Dispose();
        Assert.All(replacedResults.Concat(finalResults), result =>
            Assert.Equal(ConstructionStatus.Disposed, result.Status));
    }
}

File: examples/swift/notes-showcase/Tests/NotesShowcaseTests/NotesViewVMTests.swift
//
// NotesViewVMTests — scenario tests for NotesViewVM.
//
// Ports NotesShowcase.Tests/ViewModels/NotesViewVMTests.cs (C# Avalonia flavor).
// No conformance-ID markers (scenario IDs live in THEME-00x only).
//
// Search determinism: SearchableState debounce is bypassed by calling `search()`
// after setting `searchTerm`; this triggers the `forceSearchSubject` path, which
// fires the `filtered` sink synchronously (no scheduler involvement).
//
import XCTest
import Combine
import VMx
@testable import NotesShowcaseCore

// MARK: - Helpers

private func makeRepo(
    loadNotesDelay: TimeInterval = 0,
    saveNoteDelay: TimeInterval = 0,
    deleteNoteDelay: TimeInterval = 0
) -> InMemoryNoteRepository {
    InMemoryNoteRepository(
        seed: SeedData.build(),
        loadAllDelay: 0,
        loadNotesDelay: loadNotesDelay,
        saveNoteDelay: saveNoteDelay,
        deleteNoteDelay: deleteNoteDelay
    )
}

private func buildVM(
    repo: any NoteRepository,
    pageSize: Int = 5
) throws -> NotesViewVM {
    let hub = MessageHub()
    let dispatcher = ImmediateDispatcher.INSTANCE
    return try NotesViewVM.builder()
        .name("notes")
        .services(hub: hub, dispatcher: dispatcher)
        .repository(repo)
        .pageSize(pageSize)
        // Zero debounce + .main scheduler so debounced path resolves quickly;
        // tests drive immediate flush via `vm.search()`.
        .searchDebounce(.milliseconds(0))
        .build()
}

private actor ControlledLoadRepository: NoteRepository {
    private let seed = SeedData.build()
    private var firstLoadStarted = false
    private var startWaiters: [CheckedContinuation<Void, Never>] = []
    private var firstLoadGate: CheckedContinuation<Void, Never>?

    func waitForFirstLoad() async {
        if firstLoadStarted { return }
        await withCheckedContinuation { startWaiters.append($0) }
    }

    func releaseFirstLoad() {
        firstLoadGate?.resume()
        firstLoadGate = nil
    }

    func loadNotes(notebookId: String) async throws -> [NoteModel] {
        if notebookId == "nb-reviews" {
            firstLoadStarted = true
            let waiters = startWaiters
            startWaiters.removeAll()
            waiters.forEach { $0.resume() }
            await withCheckedContinuation { firstLoadGate = $0 }
        }
        return seed.notes.filter { $0.notebookId == notebookId }
    }

    func loadAll() async throws -> (notebooks: [NotebookModel], notes: [NoteModel]) {
        seed
    }

    func searchNotes(
        term: String,
        token: String?,
        pageSize: Int
    ) async throws -> (items: [NoteModel], nextToken: String?) {
        fatalError("not used")
    }

    func saveNote(_ note: NoteModel) async throws { fatalError("not used") }
    func deleteNote(id: String) async throws { fatalError("not used") }
    func addNotebook(_ notebook: NotebookModel) async throws { fatalError("not used") }
    func export(
        notebooks: [NotebookModel],
        notes: [NoteModel],
        path: String
    ) async throws { fatalError("not used") }
}

private func buildAndBind(
    notebookId: String = "nb-reviews",
    pageSize: Int = 5
) async throws -> (NotesViewVM, InMemoryNoteRepository) {
    let repo = makeRepo()
    let vm = try buildVM(repo: repo, pageSize: pageSize)
    try vm.construct()
    await vm.bindTo(notebookId: notebookId)
    return (vm, repo)
}

// MARK: - NotesViewVMTests

final class NotesViewVMTests: XCTestCase {

    // MARK: - bindTo

    func testBindTo_loads_notes_for_the_notebook() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        XCTAssertFalse(vm.visibleItems.isEmpty, "Expected non-empty visible items after bindTo")
        XCTAssertTrue(
            vm.filteredItems.allSatisfy { $0.model.notebookId == "nb-reviews" },
            "All filtered items should belong to nb-reviews"
        )
        XCTAssertEqual("nb-reviews", vm.boundNotebookId)
    }

    // MARK: - Paging

    func testPageSize5_yields_two_pages_for_seven_reviews_notes() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews", pageSize: 5)
        // nb-reviews has 7 notes → ceil(7/5) = 2 pages.
        XCTAssertEqual(2, vm.pageCount, "Expected 2 pages for 7 notes with page size 5")
        XCTAssertEqual(5, vm.visibleItems.count, "First page should have 5 items")
        vm.moveToNextPageCommand.execute()
        XCTAssertEqual(2, vm.visibleItems.count, "Second page should have the remaining 2 items")
    }

    func testPagination_no_op_at_boundaries() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        let first = vm.currentPageIndex
        XCTAssertFalse(vm.moveToFirstPageCommand.canExecute())
        XCTAssertFalse(vm.moveToPreviousPageCommand.canExecute())
        XCTAssertTrue(vm.moveToNextPageCommand.canExecute())
        XCTAssertTrue(vm.moveToLastPageCommand.canExecute())
        vm.moveToFirstPageCommand.execute()  // already at page 0
        XCTAssertEqual(first, vm.currentPageIndex)

        vm.moveToLastPageCommand.execute()
        let last = vm.currentPageIndex
        XCTAssertTrue(vm.moveToFirstPageCommand.canExecute())
        XCTAssertTrue(vm.moveToPreviousPageCommand.canExecute())
        XCTAssertFalse(vm.moveToNextPageCommand.canExecute())
        XCTAssertFalse(vm.moveToLastPageCommand.canExecute())
        vm.moveToNextPageCommand.execute()   // already at last page
        XCTAssertEqual(last, vm.currentPageIndex)
    }

    func testMoveToNextPage_and_back_works() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews", pageSize: 5)
        XCTAssertEqual(0, vm.currentPageIndex)
        vm.moveToNextPageCommand.execute()
        XCTAssertEqual(1, vm.currentPageIndex)
        vm.moveToPreviousPageCommand.execute()
        XCTAssertEqual(0, vm.currentPageIndex)
    }

    func testPageLabel_reflects_current_page_and_count() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews", pageSize: 5)
        XCTAssertEqual("Page 1 of 2", vm.pageLabel)
        XCTAssertEqual("Page 1 of 2", try vm.pageLabelDerived.value)
        vm.moveToNextPageCommand.execute()
        XCTAssertEqual("Page 2 of 2", vm.pageLabel)
        XCTAssertEqual("Page 2 of 2", try vm.pageLabelDerived.value)
    }

    // MARK: - Starred filter

    func testShowStarredOnly_restricts_to_starred_notes() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        vm.showStarredOnly = true
        XCTAssertTrue(
            vm.filteredItems.allSatisfy { $0.starred },
            "Only starred notes should be visible"
        )
        // nb-reviews has 2 starred notes per SeedData.
        XCTAssertFalse(vm.filteredItems.isEmpty, "Expected at least one starred note")
    }

    func testShowStarredOnly_false_restores_all_notes() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        vm.showStarredOnly = true
        let starredCount = vm.filteredItems.count
        vm.showStarredOnly = false
        XCTAssertTrue(
            vm.filteredItems.count > starredCount,
            "Disabling starred filter should show more notes"
        )
    }

    // MARK: - Custom filter predicate

    func testFilter_predicate_restricts_visible_items() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        vm.filter = { $0.title.contains("Q1") }
        XCTAssertEqual(1, vm.filteredItems.count)
        XCTAssertTrue(vm.filteredItems[0].title.contains("Q1"))
        vm.filter = nil
        XCTAssertTrue(vm.filteredItems.count > 1, "Clearing filter should restore all notes")
    }

    func testIsEmpty_true_when_filter_excludes_everything() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        XCTAssertFalse(try vm.isEmptyDerived.value)
        vm.filter = { _ in false }
        XCTAssertTrue(vm.isEmpty)
        XCTAssertTrue(try vm.isEmptyDerived.value)
        XCTAssertTrue(vm.visibleItems.isEmpty)
    }

    // MARK: - Search

    func testSearch_filters_by_title_match() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        let before = vm.filteredItems.count
        XCTAssertTrue(before > 1, "Precondition: multiple notes loaded")

        vm.searchTerm = "Q1"
        vm.search()  // force-flush past the debounce

        XCTAssertEqual(1, vm.filteredItems.count)
        XCTAssertTrue(vm.filteredItems[0].title.contains("Q1"))
    }

    func testSearch_empty_term_restores_all() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        let before = vm.filteredItems.count
        vm.searchTerm = "Q1"
        vm.search()
        XCTAssertEqual(1, vm.filteredItems.count)

        vm.searchTerm = ""
        vm.search()
        XCTAssertEqual(before, vm.filteredItems.count)
    }

    // MARK: - bindTo to a different notebook

    func testBindTo_different_notebook_swaps_items() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        let beforeIds = vm.filteredItems.map { $0.noteId }
        await vm.bindTo(notebookId: "nb-work")
        let afterIds = vm.filteredItems.map { $0.noteId }
        XCTAssertNotEqual(beforeIds, afterIds)
        XCTAssertTrue(
            vm.filteredItems.allSatisfy { $0.model.notebookId == "nb-work" },
            "All filtered items should belong to nb-work after rebind"
        )
    }

    func testOverlappingBindTo_onlyLatestRequestCanCommit() async throws {
        let repo = ControlledLoadRepository()
        let vm = try buildVM(repo: repo)
        try vm.construct()
        let transferableVM = FlagshipTransferBox(vm)

        let first = Task { [transferableVM] in
            await transferableVM.value.bindTo(notebookId: "nb-reviews")
        }
        await repo.waitForFirstLoad()
        let second = Task { [transferableVM] in
            await transferableVM.value.bindTo(notebookId: "nb-work")
        }
        await second.value
        await repo.releaseFirstLoad()
        await first.value

        XCTAssertEqual("nb-work", vm.boundNotebookId)
        XCTAssertTrue(vm.filteredItems.allSatisfy {
            $0.model.notebookId == "nb-work"
        })
    }

    // MARK: - Current selection

    func testCurrent_setter_is_idempotent() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        let first = vm.visibleItems[0]
        vm.current = first
        XCTAssertTrue(vm.current === first)
        vm.current = first  // idempotent
        XCTAssertTrue(vm.current === first)
        vm.current = nil
        XCTAssertNil(vm.current)
    }

    // MARK: - refreshNote

    func testRefreshNote_re_seats_model_and_recomputes_filter() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        // Grab the first note and apply a title filter that matches its current title.
        let original = vm.filteredItems[0].model
        vm.filter = { $0.title == original.title }
        XCTAssertEqual(1, vm.filteredItems.count)

        // Refresh with a new title — should now fall outside the filter.
        let updated = original.with(title: "RENAMED_\(original.title)")
        vm.refreshNote(updated)

        // The re-seated vm's title no longer matches the filter.
        XCTAssertTrue(vm.filteredItems.isEmpty,
                      "Updated note should not pass the old title filter")
    }

    // MARK: - Delete

    func testDeleteNote_removes_from_inner_and_clears_current() async throws {
        let repo = makeRepo(deleteNoteDelay: 0)
        let vm = try buildVM(repo: repo, pageSize: 5)
        try vm.construct()
        await vm.bindTo(notebookId: "nb-personal")

        let before = vm.inner.count
        XCTAssertTrue(before > 0, "Precondition: notes loaded")
        let target = vm.inner.at(0)
        vm.current = target

        // Await the same async relay the UI invokes through the synchronous
        // Command surface. Foreground removal completes before this returns.
        let deleteCommand = try XCTUnwrap(target.deleteCommand as? AsyncRelayCommand)
        try await deleteCommand.executeAsync()

        XCTAssertEqual(before - 1, vm.inner.count,
                       "Inner count should decrement by 1 after delete")
        XCTAssertNil(vm.current, "Current should be cleared when the selected note is deleted")
    }

    // MARK: - Reconstruct

    func testReconstruct_keeps_items_via_rebind() async throws {
        let (vm, _) = try await buildAndBind(notebookId: "nb-reviews")
        let before = vm.filteredItems.count
        try vm.reconstruct()
        // After reconstruct, prior items are torn down.
        await vm.bindTo(notebookId: "nb-reviews")
        XCTAssertEqual(before, vm.filteredItems.count,
                       "Rebind after reconstruct should restore the same item count")
    }

    // MARK: - Global search token paging

    func testRepositorySearchNotes_returnsTokenPagesOverAllNotes() async throws {
        let repo = makeRepo(loadNotesDelay: 0)

        let first = try await repo.searchNotes(term: "review", token: nil, pageSize: 2)

        XCTAssertEqual(2, first.items.count)
        XCTAssertEqual("2", first.nextToken)
        XCTAssertTrue(first.items.allSatisfy {
            "\($0.title) \($0.body) \($0.tags.joined(separator: " "))"
                .lowercased()
                .contains("review")
        })

        let second = try await repo.searchNotes(
            term: "review",
            token: first.nextToken,
            pageSize: 2
        )
        XCTAssertFalse(second.items.isEmpty)
        XCTAssertNotEqual(first.items[0].id, second.items[0].id)
    }

    func testGlobalSearchVM_refreshes_resetsTerms_andLoadsMore() async throws {
        let repo = makeRepo(loadNotesDelay: 0)
        let vm = try GlobalSearchVM.builder()
            .name("global-search")
            .services(hub: MessageHub(), dispatcher: ImmediateDispatcher.INSTANCE)
            .repository(repo)
            .pageSize(2)
            .searchDebounce(.milliseconds(0))
            .build()

        vm.searchTerm = "review"
        try await vm.refreshCommand.executeAsync()
        XCTAssertEqual(2, vm.results.count)
        XCTAssertTrue(vm.hasMore)

        try await vm.loadMoreCommand.executeAsync()
        XCTAssertGreaterThan(vm.results.count, 2)
        let replacedResults = vm.results

        vm.searchTerm = "travel"
        try await vm.refreshCommand.executeAsync()
        XCTAssertTrue(vm.results.allSatisfy { $0.model.notebookId == "nb-personal" })
        let finalResults = vm.results
        vm.dispose()
        XCTAssertTrue((replacedResults + finalResults).allSatisfy { $0.status == .disposed })
    }
}
